design/reboot-interface.md

```diff
@@ -0,0 +1,193 @@
+necessary to avoid race conditions by ensuring that the ``Online`` flag and any
+reboot requests are managed in the same place. The Host controller
+will watch RebootRequest objects and trigger a reconciliation of the linked
+Host when one appears or changes.
+
+For simplicity, we will try to avoid creating a separate controller for the
+RebootRequest at all, but instead perform all operations on it in the course of
+reconciling the BareMetalHost object that it references. However, it may yet
+prove necessary to also create a trivial controller for the purposes of basic
+housekeeping (e.g. setting the owner of the resource).
+
+The creation timestamp of the request resource will be interpreted as the time
+of the reboot request. Requests that predate the last booted time will be
+immediately marked as complete without taking any further action. The Host
+controller will timestamp all pending reboot requests as having completed the
+power off stage whenever any of the following occur:
+
+* The Host has been deprovisioned (in this case the ``lastPoweredOn`` field
+  will be empty)
+* The ``lastPoweredOn`` time is after the time of the reboot request
+* The ``poweredOn`` status is ``false``
+
+Whenever there are pending reboot requests, the Host controller will respond by
+powering off the Host (if it is not already) and timestamping all pending
+requests as having completed the power off stage. Power will then be restored
+only if/when the ``Online`` spec of the Host is true. Once Hosts are powered up
+again, any satisfied reboot requests will be timestamped as having completed
+the power on stage, with this timestamp matching the ``lastPoweredOn`` time of
+the Host.
+
+## Drawbacks
+
+This approach entails a whole new custom resource definition to maintain, and
+the instances of this resource must be managed by the Host controller, which
+adds a degree of complexity.
+
+If large numbers of requests are created but never deleted, searching for
+pending requests may start to consume significant resources.
+
+## Future Enhancements
+
+### Labelling
+
+Adding labels to partially completed and completed requests would allow us to
+exclude them when querying for pending requests. That would eliminate the
+performance concerns associated with allowing old requests to accumulate. This
+would be best performed by a RebootRequest-specific controller.
+
+### Automatic deletion
+
+The operator could automatically delete requests once they have been fulfilled
+(either immediately or after a predefined timeout). This would both eliminate
+performance concerns and reduce clutter. Clients wishing to observe data from
+the resource could set a finalizer that keeps the resource around only until
+they are done with it. This might best be performed by a RebootRequest-specific
+controller.
+
+### Custom Timestamps
+
+The resource could allow the user to specify a timestamp, which may of course
```

Comment:
It would be a good idea to more fully define the use cases for this before spending much time considering implementation.

Reply:
Yeah, I included this "future enhancements" section more as a way to record stuff we have talked about for later recollection, not as a proposal to implement it any time soon.